walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,210 @@
+            TimeRange mergeSlot = mergeSchedule.get(index1);
+            TimeRange attandeeSlot = attandeeSchedule.get(index2);
+
+            if (mergeSlot.overlaps(attandeeSlot)) {
+                TimeRange newMergeSlot = mergeTwoOverlappedTimeRanges(mergeSlot, attandeeSlot);
+                newMergeSchedule = mergeTimeRangeWithAccumulatedTimeRanges(newMergeSchedule, newMergeSlot);
+                index1++;
+                index2++;
+            }
+            else {
+                if (mergeSlot.start() < attandeeSlot.start()) {
+                    newMergeSchedule = mergeTimeRangeWithAccumulatedTimeRanges(newMergeSchedule, mergeSlot);
+                    index1++;
+                } else {
+                    newMergeSchedule = mergeTimeRangeWithAccumulatedTimeRanges(newMergeSchedule, attandeeSlot);
+                    index2++;
+                }
+            }
+
+        }
+        while (index1 < mergeSchedule.size()) {
+            TimeRange mergeSlot = mergeSchedule.get(index1);
+            newMergeSchedule = mergeTimeRangeWithAccumulatedTimeRanges(newMergeSchedule, mergeSlot);
+            index1++;
+        }
+        while (index2 < attandeeSchedule.size()) {
+            TimeRange attandeeSlot = attandeeSchedule.get(index2);
+            newMergeSchedule = mergeTimeRangeWithAccumulatedTimeRanges(newMergeSchedule, attandeeSlot);
+            index2++;
+        }
+        mergeSchedule = newMergeSchedule;
+    }
+    return mergeSchedule;
+  }
+  /* Consider if each 'missing' TimeRange in the merged schedule will be long enough to fit the meeting in
+  param: @mergeSchedule: a list of TimeRange that merges all schedule
+         @request: a MeetingRequest contains the duration of the meeting
+  return: return a list possible TimeRange that fits all given schedules
+  */
+  public static List<TimeRange> fitAllSchedules(List<TimeRange> mergeSchedule, MeetingRequest request) {
+    ArrayList<TimeRange> possibleSlots = new ArrayList<>();
+
+    if (mergeSchedule.size() == 0) {
+        if (TimeRange.END_OF_DAY - TimeRange.START_OF_DAY >= request.getDuration()) {
+            possibleSlots.add(TimeRange.WHOLE_DAY);
+        }
+    }
+    else {
+        for (int i = -1; i < mergeSchedule.size(); i++) {
+            if (i == -1) {
+                TimeRange next = mergeSchedule.get(i + 1);
+                if (next.start() - TimeRange.START_OF_DAY >= request.getDuration()) {
+                    possibleSlots.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, next.start(), false));
+                }
+                
+            }
+            else if (i == mergeSchedule.size() - 1) {
+                TimeRange curr = mergeSchedule.get(i);
+                if (TimeRange.END_OF_DAY - curr.end() >= request.getDuration()) {
+                    possibleSlots.add(TimeRange.fromStartEnd(curr.end(), TimeRange.END_OF_DAY, true));
+                }
+                
+            }
```

Comment:
If you have special cases for the beginning and end of the iteration, it might make more sense to not put them in the loop at all and shorten the iterator. For example:

```
TimeRange next = mergeSchedule.get(i + 1);
if (next.start() - TimeRange.START_OF_DAY >= request.getDuration()) {
   possibleSlots.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, next.start(), false));
}
for (int i = 0; i < mergeSchedule.size() - 1; i++) {
  TimeRange curr = mergeSchedule.get(i);
  TimeRange next = mergeSchedule.get(i + 1);
   if (next.start() - curr.end() >= request.getDuration()) {
     possibleSlots.add(TimeRange.fromStartEnd(curr.end(), next.start(), false));
   }
}
TimeRange curr = mergeSchedule.get(i);
if (TimeRange.END_OF_DAY - curr.end() >= request.getDuration()) {
  possibleSlots.add(TimeRange.fromStartEnd(curr.end(), TimeRange.END_OF_DAY, true))
}
```

Reply:
Good point @ganewman42. Did you mean this? (diffs are bloded)

```
TimeRange next = mergeSchedule.get(0);
if (next.start() - TimeRange.START_OF_DAY >= request.getDuration()) {
   possibleSlots.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, next.start(), false));
}
for (int i = 1; i < mergeSchedule.size() - 1; i++) {
  TimeRange curr = mergeSchedule.get(i);
  TimeRange next = mergeSchedule.get(i + 1);
   if (next.start() - curr.end() >= request.getDuration()) {
     possibleSlots.add(TimeRange.fromStartEnd(curr.end(), next.start(), false));
   }
}
TimeRange curr = mergeSchedule.get(mergeSchedule.size() - 1);
if (TimeRange.END_OF_DAY - curr.end() >= request.getDuration()) {
  possibleSlots.add(TimeRange.fromStartEnd(curr.end(), TimeRange.END_OF_DAY, true))
}
```

Couldn't format bold the diffs between two snippets but the diffs are:
1. '0' instead of 'i + 1'
2. 'int i = 1' instead of 'int i = 0'
3. 'mergeSchedule.size() - 1' instead of 'i'

If I'm misinterpreting your comment then please disregard this comment.